exonum-java-binding/core/rust/src/testkit/mod.rs

```diff
@@ -213,3 +214,95 @@ fn create_java_keypair<'a>(
         &[secret_key_byte_array.into(), public_key_byte_array.into()],
     )
 }
+
+// Converts Java array of `TestKitServiceInstances` to vector of `InstanceConfig`.
+fn instance_configs_from_java_array(
```

Comment:
Can't transfer all this stuff in Protobuf?

Reply:
Not sure it will simplify things WRT to the work already done. I'd better invest in polishing the existing code.

Reply:
The work already done is orthogonal to how simple things may be.

The existing code has several issues:
* inadequate local reference management given unbounded loops (not a problem in a usual use-case when there are few artifacts and services, but will cause JNI warnings when there are many of them; but I wouldn't like to establish each time it is not a problem).
* complex implementation with two nested loops — methods must be extracted for readability
* complex to review and maintain.

If we consider passing as a protobuf as a superior approach that we will replace this method with, than these issues may not be fixed (but appropriately documented in code and issues). However, with time providers, which need to be passed as references, there is no serialization option, and any issues must be resolved.

Reply:
Refactored according to comments. Decided not to switch to protobuf because of the time provider limitation.